### PR TITLE
Improving Track Information in miniAOD : 80X

### PR DESCRIPTION
**Pull request by Sam-Harper**
This is the 80X flavour version as requested by x-pog of https://github.com/cms-sw/cmssw/pull/17909

It stores trk pt and eta as well as storing all high purity tracks including those which come from electrons. For more details see the 90X PR.

The average event size increase was measured using the /RelValTTbar_13/CMSSW_8_0_21-PU25ns_80X_mcRun2_asymptotic_2016_TrancheIV_v6_Tr4GT_v6-v1/GEN-SIM-RECO ttbar relval. 

Size increase 
patPackedCandidates_packedPFCandidates__PAT. 87539.3 15788.5
patPackedCandidates_lostTracks__PAT. 278.553 121.174

to

patPackedCandidates_packedPFCandidates__PAT. 94988 15849.3
patPackedCandidates_lostTracks__PAT. 1114.2 348.642
patPackedCandidates_lostTracks_eleTracks_PAT. 201.602 90.3766

for  a total size of ~400 bytes, of which about half of which is due to <1 GeV high purity tracks been added.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@Sam-Harper @gpetruc @arizzi 
this feature looks rather common to both data and MC and will probably be not useful without MC.
Due to the scope of miniAOD v1 (data only) compared to v2 (both data and MC), I think that this PR should go only to the release used with v2.

**Comment by Sam-Harper** [comment]
@slava77  I concur. It would be most useful with a consistent set of data and MC.

File: PhysicsTools/PatAlgos/plugins/PATLostTracks.cc [hunk 1]
```diff
@@ -67,85 +108,123 @@ pat::PATLostTracks::~PATLostTracks() {}
...
+}  
 
+bool pat::PATLostTracks::passTrkCuts(const reco::Track& tr)const
+{
+  
+    bool passTrkPt = tr.pt() > minPt_;
+    bool passTrkHits = tr.numberOfValidHits() >= minHits_ && 
+                       tr.hitPattern().numberOfValidPixelHits() >= minPixelHits_;
+    bool passTrkQual = passesQuality(tr,qualsToAutoAccept_);
+    return (passTrkPt && passTrkHits) || passTrkQual;
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATLostTracks.cc]: having this method implementation more aligned with the 91X version would be nice.
cf 
```diff
+    const bool passTrkHits = tr.pt() > minPt_ && 
+                             tr.numberOfValidHits() >= minHits_ && 
+                             tr.hitPattern().numberOfValidPixelHits() >= minPixelHits_;
+    const bool passTrkQual = passesQuality(tr,qualsToAutoAccept_);
+
+    return passTrkHits || passTrkQual;
```

File: PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc [hunk 2]
```diff
@@ -190,7 +190,9 @@ void pat::PATPackedCandidateProducer::produce(edm::StreamID, edm::Event& iEvent,
 
     for(unsigned int ic=0, nc = cands->size(); ic < nc; ++ic) {
         const reco::PFCandidate &cand=(*cands)[ic];
+	float trkPt = cand.pt();
         float phiAtVtx = cand.phi();
+        float etaAtVtx = cand.eta();
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc]: please keep in sync with 91X where trkPt and etaAtVtx are defined/initialized later [L220, 222 in this version]

**Comment by slava77** [comment]
some fixes are needed here before I can sign

**Comment by Sam-Harper** [comment]
@slava77, what I'll do is resync it completely with the 91X version. 80X and 9X had diverged by the time I added my fixes. I'll ensure that any changes are cosmetic and then just replace 80X with 9X.  Would that be a good solution? 

Also with regards to your comment, I fully agree this should only go in when we are planning to remake both MC and data miniAOD, given that it changes the returned track pt (although in a way that is recoverable with a hack and only for a small number of events). Let me know how you want to proceed. 

Thanks

**Comment by arizzi** [comment]
@Sam-Harper is it going to be backward compatible? (i.e. I assume that if deltaPt is zero it behaves as the old code)

**Comment by Sam-Harper** [comment]
@arizzi Yes. When reading old miniAOD, it'll give the old results. Its more you second you produce anything with this, it becomes different (so if you only do data, you are slightly inconsistent with MC) but in a way that probably nobody cares about. 

However just to double check, I'll run some tests to ensure that when reading, it returns exactly the same values.

**Comment by slava77** [comment]
On 3/27/17 2:08 AM, Sam-Harper wrote:
> @slava77 <https://github.com/slava77>, what I'll do is resync it
> completely with the 91X version. 80X and 9X had diverged by the time I
> added my fixes. I'll ensure that any changes are cosmetic and then just
> replace 80X with 9X. Would that be a good solution?
> 

sounds good.

> Also with regards to your comment, I fully agree this should only go in
> when we are planning to remake both MC and data miniAOD, given that it
> changes the returned track pt (although in a way that is recoverable
> with a hack and only for a small number of events). Let me know how you
> want to proceed.

Once the PR is updated and looks good, I intend to sign it and put it on
hold until it's ready to merge for miniAOD v2.

**Comment by slava77** [comment]
@Sam-Harper 
please clarify if you intend to update this soon.
There is still enough time until v2, but it may be better to address pending issues already.

**Comment by Sam-Harper** [comment]
Yes. I'm on a top priority critical task for the E/gamma HLT pixel matching which needs my full attention right now. It should resolve to a point its no longer critical in a day or two. This is second on the list.

**Comment by arizzi** [comment]
@slava77 if we are going to do v2 of miniaod with 9XY this is not needed.
If we do v2 of miniaod with 80X we have ~10 other PRs to back port

**Comment by slava77** [comment]
On 5/2/17 8:13 AM, arizzi wrote:
> @slava77 <https://github.com/slava77> if we are going to do v2 of
> miniaod with 9XY this is not needed.
> If we do v2 of miniaod with 80X we have ~10 other PRs to back port

that's fine.
This PR wasn't ported fully consistently.
I wanted to get the pending review issues resolved and have it ready to
go in the 80X v2 happens.
Better do it now than when all other backports need to happen.

**Comment by Sam-Harper** [comment]
its fine, I dont mind doing it. It sadly though taken second fiddle to other more critical issues. Its now bubbled near the top of my todo list now the pixel matching and digi filtering PRs are done, with the pixel matching PR being absolutely critical for CMS data taking and needed urgently, hence working full time on that.

**Comment by slava77** [comment]
ping
just a weekly check of the heartbeat.

**Comment by slava77** [comment]
-1

updates are expected since a while. Apparently this is now too low of a priority.

**Pull request closed without merging**